commands/new_function_test.go

```diff
@@ -0,0 +1,171 @@
+		expectedMsg: InvalidYAMLMap,
+	},
+	{
+		title:       "new_append_invalid_3",
+		funcName:    "new-test-append-invalid-3",
+		funcLang:    "python3",
+		file:        "invalid3.yml",
+		expectedMsg: InvalidYAMLMsg,
+	},
+}
+
+func runNewFunctionTest(t *testing.T, nft NewFunctionTest) {
+	funcName := nft.funcName
+	funcLang := nft.funcLang
+	var funcYAML string
+
+	originalYAMLFile := "new_test.orig_yaml"
+	if len(nft.file) == 0 {
+		funcYAML = funcName + ".yml"
+	} else {
+		funcYAML = nft.file
+		test.Copy(funcYAML, originalYAMLFile)
+	}
+
+	// Cleanup the created directory
+	defer func() {
+		os.RemoveAll(funcName)
+		if len(nft.file) == 0 {
+			os.Remove(funcYAML)
+		} else {
+			test.Copy(originalYAMLFile, funcYAML)
+			os.Remove(originalYAMLFile)
+		}
+	}()
+
+	cmdParameters := []string{
+		"new",
+		"--name=" + funcName,
+		"--lang=" + funcLang,
+	}
+	if len(nft.file) > 0 {
+		cmdParameters = append(cmdParameters, "--yaml="+nft.file)
+	}
+
+	stdOut := test.CaptureStdout(func() {
+		faasCmd.SetArgs(cmdParameters)
+		faasCmd.Execute()
+	})
+
+	// Validate new function output
+	if found, err := regexp.MatchString(nft.expectedMsg, stdOut); err != nil || !found {
+		t.Fatalf("Output is not as expected:\n%s", stdOut)
+	}
+
+	if nft.expectedMsg == SuccessMsg {
+
+		// Make sure that the folder and file was created:
+		if _, err := os.Stat("./" + funcName); os.IsNotExist(err) {
+			t.Fatalf("%s/ directory was not created", funcName)
+		}
+
+		if _, err := os.Stat(funcYAML); os.IsNotExist(err) {
+			t.Fatalf("\"%s\" yaml file was not created", funcYAML)
+		}
+
+		// Make sure that the information in the YAML file is correct:
+		parsedServices, err := stack.ParseYAMLFile(funcYAML, "", "")
+		if err != nil {
+			t.Fatalf("Couldn't open modified YAML file \"%s\" due to error: %v", funcYAML, err)
+		}
+		services := *parsedServices
+
+		var testServices stack.Services
+		testServices.Provider = stack.Provider{Name: "faas", GatewayURL: "http://localhost:8080"}
+		if !reflect.DeepEqual(services.Provider, testServices.Provider) {
+			t.Fatalf("YAML `provider` section was not created correctly for file %s: got %v", funcYAML, services.Provider)
+		}
+
+		testServices.Functions = make(map[string]stack.Function)
+		testServices.Functions[funcName] = stack.Function{Language: funcLang, Image: funcName, Handler: "./" + funcName}
+		if !reflect.DeepEqual(services.Functions[funcName], testServices.Functions[funcName]) {
+			t.Fatalf("YAML `functions` section was not created correctly for file %s, got %v", funcYAML, services.Functions[funcName])
+		}
+	}
+
+}
+
+func Test_newFunctionTests(t *testing.T) {
+	defer func() {
+		os.Remove("template")
+	}()
+
+	// Change directory to testdata
+	if err := os.Chdir("testdata/new_function"); err != nil {
+		t.Fatalf("Error on cd to testdata dir: %v", err)
+	}
+
+	// Symlink template directory at root to current directory to avoid re-downloading templates
+	os.Symlink("../../../template", "template")
```

Comment:
I think this could be asking for trouble. Can we do a `cp` or just generate a minimum set of files and have that as part of the tests?

Reply:
Sure, I'll copy the folder over instead of using `os.Symlink`.